Filter classified listings by published status for mods

citizen428 opened the pull request:
## What type of PR is this?

- [X] Refactor
- [X] Feature

## Description

This PR adds filtering by published status to `/internal/listings`. Unpublished listings will be filtered out by default.

## Related Tickets & Documents

Closes https://github.com/thepracticaldev/tech-private/issues/424

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/47985/79962618-77f7e000-84b2-11ea-88fb-3cf7b5ba71b6.gif)

## Added tests?

- [X] yes

## Added to documentation?
- [x] no documentation needed

citizen428 commented on app/controllers/internal/classified_listings_controller.rb:
```diff
@@ -18,14 +19,14 @@ def update
     clear_listings_cache
     flash[:success] = "Listing updated successfully"
-    redirect_to "/internal/listings/#{@classified_listing.id}/edit"
+    redirect_to edit_internal_listing_path(id: @classified_listing.id)
```
Since I was here already...

citizen428 commented on app/controllers/internal/classified_listings_controller.rb:
```diff
@@ -18,14 +19,14 @@ def update
     @classified_listing.destroy
     flash[:warning] = "'#{@classified_listing.title}' was destroyed successfully"
-    redirect_to "/internal/listings"
+    redirect_to internal_listings_path
```
Same as above.

citizen428 commented on app/views/internal/classified_listings/index.html.erb:
```diff
@@ -11,7 +11,12 @@
         <div class="form-group">
           <%= label_tag(:filter, "Category") %>
-          <%= select_tag(:filter, options_for_select([""] + ClassifiedListing.categories_available.keys, params[:filter])) %>
+          <%= select_tag(:filter, options_for_select(ClassifiedListing.categories_available.keys, params[:filter]), include_blank: true) %>
```
Changed this to use the more standard `include_blank` option.

citizen428 commented on app/views/internal/classified_listings/index.html.erb:
```diff
@@ -11,7 +11,12 @@
+        </div>
+        <div class="form-group">
+          <%= label_tag(:published, "Published only") %>
+          <%= hidden_field_tag(:published, 0) %>
```
Unlike `f.check_box` `check_box_tag` does not take care of this by itself, so we have to add the unchecked option ourselves via `hidden_tag`.

citizen428 commented on app/controllers/internal/classified_listings_controller.rb:
```diff
@@ -4,6 +4,7 @@ class Internal::ClassifiedListingsController < Internal::ApplicationController
 
   def index
     @classified_listings = ClassifiedListing.includes(%i[user organization]).page(params[:page]).order("bumped_at DESC").per(50)
+    @classified_listings = @classified_listings.published unless include_unpublished?
```
@jacobherrington Do we actually use the search box of the form anywhere?

lightalloy commented on app/views/internal/classified_listings/index.html.erb:
```diff
@@ -11,7 +11,12 @@
+        <div class="form-group">
+          <%= label_tag(:published, "Published only") %>
+          <%= hidden_field_tag(:published, 0) %>
+          <%= check_box_tag(:published, 1, true) %>
```
If I uncheck the checkbox and click "filter", the checkbox will be checked after the filtering. I would expect it to remain unchecked after unchecking. How do you feel about making it checked or unchecked based on the corresponding param (when it exists)?

citizen428 replied:
The reason I did that was as a workflow optimization for @michael-tharrington. My rationale:

1. Most of the time he wants to see the filtered list.
2. If he really wants to see all listings he can uncheck the box and hit "filter".
3. If he wants to get back to the filtered list, he only needs to press "filter" instead of fiddling with the check box. 

I see how it's confusing though, so will change it.

msarit commented:
I second what Anna said about the `Published only` checkbox bug. Here is a brief video with my observations:

https://www.loom.com/share/198b5444f1ab4465afbd1461e1fcb2cf

msarit commented on app/views/internal/classified_listings/index.html.erb:
```diff
@@ -11,7 +11,12 @@
+        <div class="form-group">
+          <%= label_tag(:published, "Published only") %>
+          <%= hidden_field_tag(:published, 0) %>
+          <%= check_box_tag(:published, 1, true) %>
```
I second what Anna has flagged re: `Published only` checkbox bug; I captured my observations in the video below - hope it's helpful! 🙏 

https://www.loom.com/share/198b5444f1ab4465afbd1461e1fcb2cf

citizen428 replied:
1. What the checkbox is supposed to do is to toggle between all listings (including published "no") and only the ones with a published status of "yes". It should never only show unpublished listings. Maybe the label of the checkbox is not great, so I'm open to suggestions here  😃 

2. The checkbox behavior was somewhat intentional, as this is an internal tool with one main user (@michael-tharrington) and this is going to save him one click. Given that it confused two people though I'll change it to behave more like a traditional checkbox.

BTW, @lightalloy was only referencing point 2. above in her comment, not point 1.

msarit replied:
ahhhhh ok I understand. Yes the checkbox label didnt make it very clear to me. Thanks for explaining.

citizen428 commented:
Thanks for the feedback @lightalloy and @msarit. Based on that I made the following changes:

1. Changed the label to "Include unpublished listings".
2. Changed the params name to `include_unpublished`.
3. Flipped the logic around accordingly.
4. Make the checkbox respect the param.
5. Resolve merge conflicts caused by merging #7250.

Here's a recording of how it works now, the listing names illustrate the intended behavior:

![2020-04-23 11 19 45](https://user-images.githubusercontent.com/47985/80058877-a0381b00-8554-11ea-95a5-401ef490df9f.gif)
